### DISCUSSION TITLE
Medplum Server Plugin Architecture

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#1424] [Ideas]
# Medplum Server Plugin Architecture

The Medplum API server is currently a single NodeJS project, which can be run on bare Linux or in AWS.  This is good for simplicity, but has a few limitations:

- AWS dependencies are required even if not running on AWS
- Sub-services are not clearly modular, and hard coded checks against config settings
- No clear path to adding alternate sub-service implementations

To be clear, this is **not** a discussion about microservices.  We are committed to a single server deployment model.  This is purely a code architecture discussion.

Examples of "sub services":

- **Secrets / config management**
  - Currently assumed to be either a JSON config file or AWS Parameter Store + AWS Secrets Manager
  - We have active customer requests for alternate providers:
    - HashiCorp Vault
    - Azure Key Vault
    - Google Secrets Manager
- **Logging**
  - Currently assumed to be stdout
  - On AWS Fargate, stdout is redirected to AWS CloudWatch Logs
  - Active customer requests for alternate providers:
    - Splunk
    - Sumo Logic
    - Azure Monitor Logs
    - Google Cloud Logging
- **Email**
  - Currently only support AWS SES, even on bare Linux
  - This makes it difficult to test on localhost, and typically requires giving developers an account with SES permissions
  - Using `nodemailer` as a common interface would be better, to support a variety of options, including mock logging
  - Active customer requests for alternate providers:
    - SMTP
    - Twilio SendGrid
    - Azure Communication Services
- **Binary file storage**
  - Currently support (1) local file system or (2) AWS S3 + CloudFront
  - There is extra complexity here for how we use "Presigned URLs" for better client-side performance
  - In a strict FHIR server, this content would only be available using the FHIR `Binary` API endpoints
  - However, that requires forms of authentication that can interfere with client-side performance
    - For example, it's not possible to stream video files, and instead you need to download the entire video
    - That completely breaks large (i.e., 100MB+) video experience
  - Alternative providers:
    - Microsoft Azure Blob
    - Google Cloud Storage
- **Bot execution**
  - Currently only support AWS Lambda
  - In the past, we supported running bots in-process using `vm.createContext()`
    - This is great for developer experience, but generally not acceptable for a multi tenant architecture
    - It could still be a good option for self-hosted single tenant customers, who completely control the environment
  - Active customer requests for alternate providers:
    - NodeJS VM Context
    - Azure Automation
    - Google Cloud Functions

## Proposal

The package currently known as `@medplum/server`  would get chopped up into a series of component packages.  While `@medplum/server` is currently **not** published in npm, all of these component packages **would** be published.

- **`@medplum/server-plugins`**
  - Define interfaces for all of the sub services
  - Maybe some plugin helpers?
  - Maybe a dependency on `@medplum/core`?
  - Minimal dependencies
  - No dependency on `@medplum/server-base`
- **`@medplum/server-base`**
  - Common server implementation
  - Dependency on `express` and related express plugins
  - All routing and auth endpoints
- **`@medplum/server-config-aws`**
  - Current implementation of config settings using AWS Parameter Store and AWS Secrets Manager
- **`@medplum/server-logging-aws-cloudwatch`**
  - Current implementation of logging using AWS CloudWatch
- **`@medplum/server-logging-aws-ses`**
  - Current implementation of email using AWS SES
- **`@medplum/server-logging-aws-s3`**
  - Current implementation of storage using AWS S3
- **`@medplum/server-bot-runner-aws-lambda`**
  - Current implementation of bot execution using AWS Lambda

Then, a fully configured server would do the following:

- Dependency on `@medplum/server-base`
- Dependency on plugin implementations
- Very short `index.ts` that puts them together

For example, an AWS server:

```ts
import { MedplumServer } from '@medplum/server-base';
import { MedplumAwsLambdaBotRunner } from '@medplum/server-bot-runner-aws-lambda';
import { MedplumAwsConfig } from '@medplum/server-config-aws';
import { MedplumAwsSesEmail } from '@medplum/server-email-aws-ses';
import { MedplumAwsCloudWatchLogging } from '@medplum/server-logging-aws-cloudwatch';
import { MedplumAwsS3Storage } from '@medplum/server-storage-aws-s3';
import express from 'express';

const app = express();
const port = 3000;

const medplumServer = new MedplumServer({
  config: new MedplumAwsConfig(),
  logging: new MedplumAwsCloudWatchLogging(),
  email: new MedplumAwsSesEmail(),
  storage: new MedplumAwsS3Storage(),
  botRunner: new MedplumAwsLambdaBotRunner(),
});

app.use('/', medplumServer);

app.listen(port, () => {
  console.log(`Medplum server listening on port ${port}`);
});
```

As a convenience, we could publish default server configurations as Docker images.

**Comment by reshmakh** [comment]
This seems very desirable from a cross platform perspective, and BYO logging, email, storage and secrets seem like the right abstractions to expose.  Potentially including plugins for things like linked data warehouse, or even other datastores seem desirable as well.

I believe that is would be possible to abstract the server to support plugins and that would be great. 

Is it possible to get a unified `bot` development experience?  Secrets and email across clouds appear to have different interfaces.

**Reply by codyebberson** [reply]
> Is it possible to get a unified bot development experience?

This is a good question, and there are layers to the response.

To the direct point of "can we provide an abstraction between bots and email/secrets?", I think the answer is "yes, we already do!"

Email:  The `MedplumClient` class has a `sendEmail` method, which posts to a Medplum API endpoint `/email/v1/send`.  The interface is  [nodemailer](https://nodemailer.com/extras/mailcomposer/), which is a thin wrapper around the RFC822 format standard. That should cleanly map to any underlying email provider.

Secrets:  We currently pass in Medplum secrets to the bot in the `handler()`.  That's not quite 1:1 with what you were asking, but I think it's a good alternative, which again should map to any underlying execution framework.

There's another layer to the question regarding "development experience".  For example, consider the dev experience of raw AWS Lambda vs SST, which includes niceties such as proxying to dev machine so you can set a breakpoint.  All of the various execution framework options have various convenience tools for localhost development, and we should be thoughtful about how to expose/leverage those advantages.

**Comment by rahul1** [comment]
I agree with @reshmakh that this is a great step in the right direction. This should make it easier to: support additional cloud providers, plug in additional service providers, and enable community contributions. 

I had a few questions / comments about some of the implementation details. **I would love to get community feedback on these as well**

1. Would it be possible and/or desirable to move the server config from `index.ts`, as you propose, to a config file? Thinking something like: 

```
{
...
"bot-runner": "@medplum/server-bot-runner-aws-lambda",
"config-provider":  "@medplum/server-config-aws",
"email-service": "@medplum/server-email-aws-ses"
...
}
```
To me it seems like writing such a simple `index.ts` for every instance introduces boilerplate, but I'm not sure if this just introduces config hell. 

2. How should we manage configuration parameters for each module (i.e. what we currently do in `medplum.config.json`? Do we want to assume that all modules that implement the same function (e.g all the implementations of the email module) share the same configuration schema? Or is it possible that, for instance, the SMTP implementation of email needs different config settings than the SES implementation? Should each module implementation define their own configuration parameters? I'm not sure what the right answer is here

**Reply by codyebberson** [reply]
@rahul1  Good questions

> Would it be possible and/or desirable to move the server config to a config file? 

Possible: Maybe?

- Implementing packages would have to follow a strict convention for how exports work
- The server implementation would be a complex, because you'd be doing some pretty abstract/meta programming
- Ironically, it would be straightforward with CommonJS, because `require()` is just a function.  It would be comparatively more complicated with ESM, because you'd need to do dynamic imports.

Desirable?

- There might be some benefits to config driven / data driven, but I think this would introduce more complexity in the `base-server` package
- It would put more burden on plugin implementers to get package exports exactly correct
- It might be a missed opportunity to passing parameters and other config options into the plugins themselves?

> How should we manage configuration parameters for each module?

This is a good question, and probably requires some research.

The current `MedplumServerConfig` assumes a somewhat flat key/value pair model.  There are exceptions (namely, Postgres and Redis config settings).  We could embrace those exceptions, and say "everything is namespaced by  the plugin name"?

> Should each module implementation define their own configuration parameters? 

That's probably the only sane way to do this.
